src/test/java/htsjdk/variant/bcf2/BCF2WriterUnitTest.java

```diff
@@ -0,0 +1,253 @@
+    private Set<VCFHeaderLine> metaData;
+
+    private Set<String> additionalColumns;
+
+    private File tempDir;
+
+    /**
+     * create a fake header of known quantity
+     *
+     * @param metaData          the header lines
+     * @param additionalColumns the additional column names
+     * @return a fake VCF header
+     */
+    public static VCFHeader createFakeHeader(final Set<VCFHeaderLine> metaData, final Set<String> additionalColumns,
+                                             final SAMSequenceDictionary sequenceDict) {
+        metaData.add(new VCFHeaderLine(VCFHeaderVersion.VCF4_0.getFormatString(),
+                                       VCFHeaderVersion.VCF4_0.getVersionString()));
+        metaData.add(new VCFHeaderLine("two", "2"));
+        additionalColumns.add("extra1");
+        additionalColumns.add("extra2");
+        final VCFHeader ret = new VCFHeader(metaData, additionalColumns);
+        ret.addMetaDataLine(new VCFInfoHeaderLine("GT", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFInfoHeaderLine("DP", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFInfoHeaderLine("BB", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFInfoHeaderLine("GQ", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("GT", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("DP", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("BB", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("GQ", 1, VCFHeaderLineType.String, "x"));
+        ret.setSequenceDictionary(sequenceDict);
+        return ret;
+    }
+
+    @BeforeClass
+    private void createTemporaryDirectory() {
+        tempDir = TestUtil.getTempDirectory("BCFWriter", "StaleIndex");
+    }
+
+    @AfterClass
+    private void deleteTemporaryDirectory() {
+        for (File f : tempDir.listFiles()) {
+            f.delete();
+        }
+        tempDir.delete();
+    }
+
+    /**
+     * test, using the writer and reader, that we can output and input BCF without problems
+     */
+    @Test
+    public void testWriteAndReadBCF() throws IOException {
+        final File fakeVCFFile = VariantBaseTest.createTempFile("testWriteAndReadVCFBody.", ".bcf");
+
+        Tribble.indexFile(fakeVCFFile).deleteOnExit();
+        metaData = new HashSet<VCFHeaderLine>();
+        additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+        try (final VariantContextWriter writer = new VariantContextWriterBuilder()
+                .setOutputFile(fakeVCFFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.ALLOW_MISSING_FIELDS_IN_HEADER, Options.INDEX_ON_THE_FLY))
```

Comment:
BCF requires all fields to be present in the header. I'm a bit surprised this isn't rejected, but I think it should be removed for this test.